serde_derive/src/ser.rs

```diff
@@ -877,20 +924,33 @@ fn serialize_struct_visitor(
             }
 
             let span = Span::def_site().located_at(field.original.span());
-            let func = struct_trait.serialize_field(span);
-            let ser = quote! {
-                try!(#func(&mut __serde_state, #key_expr, #field_expr));
+            let ser = if field.attrs.flatten() {
+                quote! {
+                    try!((#field_expr).serialize(_serde::private::ser::FlatMapSerializer(&mut __serde_state)));
```

Comment:
Please use fully qualified [UFCS](https://doc.rust-lang.org/book/first-edition/ufcs.html) for the `serialize` method call. Otherwise this is screwed up by other traits the user may have in scope.